hibernate-core/src/main/java/org/hibernate/type/AbstractStandardBasicType.java

```diff
@@ -349,8 +345,19 @@ public final Type getSemiResolvedType(SessionFactoryImplementor factory) {
 	}
 
 	@Override
-	@SuppressWarnings({ "unchecked" })
 	public final Object replace(Object original, Object target, SharedSessionContractImplementor session, Object owner, Map copyCache) {
+		if ( original == null && target == null ) {
+			return null;
+		}
+
+		if ( original == LazyPropertyInitializer.UNFETCHED_PROPERTY ) {
+			return target;
+		}
+
+		if ( target == LazyPropertyInitializer.UNFETCHED_PROPERTY ) {
+			return getReplacement( (T) original, null, session );
+		}
```

Comment:
You might want to have a look at the other `replace()` method below, which also calls `getReplacement`, but without executing this code. Probably a bug?

It might be a dumb suggestion, but couldn't we just move all this code to the caller(s) of `replace()`? Then any type would benefit from this code, not just types extending `AbstractStandardBasicType`. From what I can see it's rather simple, and it would remove the need for type implementations to refer to `LazyPropertyInitializer`, which is an SPI type...